### ISSUE TITLE
RequireThis: Incorrectly triggers for record fields in compact constructor

### ISSUE THREAD

**Issue by zml2008** [checkstyle/checkstyle#8873]
Check documentation: https://checkstyle.org/config_coding.html#RequireThis

```
/var/tmp $ javac --enable-preview --release 14 RequireThisRecord.java
Note: RequireThisRecord.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
```

`/var/tmp $ cat checkstyle-recordthis.xml`
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <property name="severity" value="${severity}" default="warning"/>

    <property name="fileExtensions" value="java, properties, xml"/>
    <!-- Excludes all 'module-info.java' files              -->
    <!-- See https://checkstyle.org/config_filefilters.html -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$"/>
    </module>

    <module name="TreeWalker">
        <module name="RequireThis">
            <property name="validateOnlyOverlapping" value="false"/>
        </module>
    </module>
</module>
```

`/var/tmp $ cat RequireThisRecord.java`
```java
import java.util.Map;

record RequireThisRecord(String name, Map<String, String> items) {

    public RequireThisRecord {
        if (name == null) {
            name = "<unknown>";
        }
        items = Map.copyOf(items);
    }

}
```

```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-8.36.2-all.jar -c checkstyle-recordthis.xml RequireThisRecord.java
Starting audit...
[WARN] RequireThisRecord.java:6:13: Reference to instance variable 'name' needs "this.". [RequireThis]
[WARN] RequireThisRecord.java:7:13: Reference to instance variable 'name' needs "this.". [RequireThis]
[WARN] RequireThisRecord.java:9:9: Reference to instance variable 'items' needs "this.". [RequireThis]
[WARN] RequireThisRecord.java:9:28: Reference to instance variable 'items' needs "this.". [RequireThis]
Audit done.

```
---------------

I'm back again :) been playing with records some more in Configurate, and came across this issue. The `RequireThis` check seems to treat accesses in the record constructor as acting on the backing fields, not the (implicit) constructor parameters, so it expects them to be qualified with `this.`. However, adding `this.` to the variables results in a compile-time error, so this check should not trigger an error.

From [the JLS section on records](https://docs.oracle.com/javase/specs/jls/se15/preview/specs/records-jls.html#jls-8.10), section 8.10.4:

> It is a compile-time error if an assignment occurs (16) to a field corresponding to a record component of the record class in the body of the compact constructor.
>
> All fields corresponding to the record components of the record class are implicitly initialized to the value of the corresponding formal parameter after the body of the compact constructor. These fields are implicitly initialized in the order that they are declared in the record component list.

I came across this as part of some work I'm doing locally to validate the API -- the full original class that caused the error is at https://gist.github.com/zml2008/5f6de6484eed5b9d8451c5e1625f8572

**Comment by pbludov** [comment]
Interesting that
```java
import java.util.Map;

record RequireThisRecord(String name, Map<String, String> items) {

    public RequireThisRecord {
        if (name == null) {
            name = "<unknown>";
        }
        this.items = Map.copyOf(items);
    }

}
```
compiles fine. In the other three places, use of `this.` causes compilation to fail.

**Comment by pbludov** [comment]
This compiles too:
```java
import java.util.Map;
import java.util.Optional;

record RequireThisRecord(String name, Map<String, String> items) {

    public RequireThisRecord {
        this.name = Optional.of(name).orElse("<unknown>");
        this.items = Map.copyOf(items);
    }

}
```
This explains why all tests passed in #8473

**Comment by zml2008** [comment]
What Java version are you testing on? I believe some of the requirements for compact constructors were tightened up in JDK 15 -- at least according to the text in the final JEP issue: https://bugs.openjdk.java.net/browse/JDK-8246771

Testing locally, the examples compile fine on Java 14 but not on Java 15, so the check probably got caught by a change in spec?

**Comment by pbludov** [comment]
> What Java version are you testing on?

Java 14. We have Java 15 CI for the latest master: https://cirrus-ci.com/task/4594812944908288
But there are no "correct" cases in the tests, only violations are [here:](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisRecordAsTopLevel.java)

**Comment by anhminhtran235** [comment]
I am on it

**Comment by anhminhtran235** [comment]
@zml2008 @pbludov  Sorry for the delay.

```
$cat RequireThisRecord.java
import java.util.Map;

record RequireThisRecord(String name, Map<String, String> items) {

    public RequireThisRecord {
        if (name == null) {
            this.name = "<unknown>";
        } else {
            this.name = name;
        }
        this.items = Map.copyOf(items);
    }

}
```

Compile with Java 14.0.1: Compile successful
Compile with Java 15.0.1:
```
RequireThisRecord.java:7: error: cannot assign a value to final variable name
            this.name = "<unknown>";
                ^
RequireThisRecord.java:9: error: cannot assign a value to final variable name
            this.name = name;
                ^
RequireThisRecord.java:11: error: cannot assign a value to final variable items
        this.items = Map.copyOf(items);
            ^
Note: RequireThisRecord.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
3 errors
```

2 java versions give different compiling results. 
So that means we need to have different logic for each version?

**Comment by romani** [comment]
No, it is mandatory for sources to compile.
If jdk15 give compilation error, it means that bug is not valid. It is very rare case. 
We will not support intermediate state of compiler.

**Comment by zml2008** [comment]
Sorry, I think there is some misunderstanding @romani. My issue is that using `this` was allowed in the Java 14 record preview but is no longer allowed in Java 15's second preview of records. Checkstyle's existing behaviour is correct for Java 14, but not for Java 15 or [what will be the final records in Java 16](https://openjdk.java.net/jeps/395).

See the original source file I posted -- it is valid in both Java 14 and Java 15, but the addition of `this` as proposed by the `RequireThis` check is only valid in Java 14, not in any later versions -- so Checkstyle's current state is supporting an intermediate state of the compiler. That's what I'd like fixed.

At least for me, the point of supporting records is to do early testing to be ready for when records are no longer a preview feature, which means targeting the latest preview variant. Especially in this case, where everything valid in Java 15 is valid in Java 14 but not the other way around, what Java 14 does should not guide development.

**Comment by romani** [comment]
>  but the addition of this as proposed by the RequireThis check is only valid in Java 14, not in any later versions

sorry, you are right, I misread issue and comments.

**Comment by anhminhtran235** [comment]
@romani @zml2008 If I understand correctly, in Java 15 `this` is not allowed in record's constructor at all?
```
// cannot use this.s
record RequireThisRecord(String s) {
    public RequireThisRecord {
        System.out.println(this.s); // compile error: variable s might not have been initialized
    }
}
```
```
// cannot assign value to this.s
record RequireThisRecord(String s) {
    public RequireThisRecord {
        this.s = "a"; // compile error:  cannot assign a value to final variable s
    }
}
```
So `RequireThis` should be disabled in record's constructor. But in other place, `this` can still be used such as:
```
// compiles fine
record RequireThisRecord(String s) {
    public void foo(String s) {
        System.out.println(this.s);
        System.out.println(s);
        s = this.s;
    }
}
```
That means we should still `require this` in other places in records (and also take into account that assigning to this.variable is not possible)?
```
record RequireThisRecord(String s) {
    public void foo(String s) {
        this.s = "a"; // compile error: cannot assign a value to final variable s
        s = this.s;  // ok
        s = "b"; // ok
    }
}
```

**Comment by romani** [comment]
we should skip in record c-tors for only, all other method should be under validation.

**Comment by anhminhtran235** [comment]
Ok. I will work on that. Please allow me extra time as this is currently a busy period for me.

**Comment by anhminhtran235** [comment]
@romani @zml2008  Sorry for the delay. I found out 1 more bug:

```
$ cat RequireThisRecord.java
record RequireThisRecord(String name) {
    public void foo(String name) {
        name = name;
    }
}

$ cat config.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="RequireThis">
            <property name="validateOnlyOverlapping" value="false"/>
        </module>
    </module>
</module>

javac --source 15 --enable-preview RequireThisRecord.java // java 14 and 15 compiles fine

java -jar .\checkstyle-8.37-all.jar -c config.xml RequireThisRecord.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing RequireThisRecord.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:311)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:221)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:408)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:331)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
Caused by: java.lang.NullPointerException: Cannot invoke "com.puppycrawl.tools.checkstyle.api.DetailAST.findFirstToken(int)" because "mods" is null
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck$ClassFrame.hasFinalField(RequireThisCheck.java:1513)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.canAssignValueToClassField(RequireThisCheck.java:903)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.getClassFrameWhereViolationIsFound(RequireThisCheck.java:682)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.getFieldWithoutThis(RequireThisCheck.java:507)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.processIdent(RequireThisCheck.java:460)
        at com.puppycrawl.tools.checkstyle.checks.coding.RequireThisCheck.visitToken(RequireThisCheck.java:394)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:340)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:451)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:278)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:151)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:87)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:337)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:298)
        ... 5 more
Checkstyle ends with 1 errors.
```
Should I fix both bugs in the same PR, or should I create a separate issue for the new bug?

**Comment by romani** [comment]
We create new issue only if fixing of it requires a lot other unrelated changes. If code changes are around the same code, ok to fix both issues in scope of one

**Comment by anhminhtran235** [comment]
When logging out the error:

https://github.com/checkstyle/checkstyle/blob/47a6dc8c0b9b81e5e83710080225dcb0aa788e9b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java#L459-L464

Checking first if the `ast` is in `COMPACT_CTOR_DEF` solves the current issue.

```
 if (checkFields) {
     final AbstractFrame frame = getFieldWithoutThis(ast, parentType);
     final boolean canUseThis = !isInCompactConstructor(ast); // In java 15, 'this' cannot be used in COMPACT_CTOR_DEF
     if (frame != null && canUseThis) {
         logViolation(MSG_VARIABLE, ast, frame);
     }
 }
```

**Comment by anhminhtran235** [comment]
The change in Java 15 causes another (small) error in our check too:
```
$ cat InputRequireThis.java
public record InputRequireThis(int x, int y) {
  InputRequireThis(int x) {
    this(x, 42);
    x = x; // wrong error message
  }
}

java -jar .\checkstyle-8.37-all.jar -c config.xml InputRequireThis.java
Starting audit...
[ERROR] E:\checkstyle-playground\InputRequireThis.java:4:5: Reference to instance variable 'x' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 1 errors.
```

Because 'x' is `final` instance variable, `this.x = x` is not possible. However, the error is at 4:5 (The first 'x'), while it should be at 4:9 (The second 'x').

**Comment by anhminhtran235** [comment]
> Because 'x' is final instance variable, this.x = x is not possible. However, the error is at 4:5 (The first 'x'), while it should be at 4:9 (The second 'x').

@romani  I'm working on this one. I guess this should be in this PR too?

**Comment by romani** [comment]
>I guess this should be in this PR too? 

Up to you. We can split in few PRs.

**Comment by anhminhtran235** [comment]
It seems that RequireThis hasn't handled the case where an instance variable is final:
```
$ cat Test.java
public class Test {
  final int x = 0;
  public Test(int x) {
    x = x;
  }
}

java -jar .\checkstyle-8.37-all.jar -c config.xml Test.java
Starting audit...
[ERROR] E:\checkstyle-playground\Test.java:5:5: Reference to instance variable 'x' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 1 errors.
```

Similar to [my previous comment](https://github.com/checkstyle/checkstyle/issues/8873#issuecomment-723727095), error should occur at 5:9 since `this.x = x` gives a compile error (`x` is `final`).
I think this problem is more about `RequireThis` for `final variable` than about `compact constructor`. I'm gonna make a commit to fix `compact constructor` problem in this issue, and then open another issue for `final variable`.

**Comment by romani** [comment]
Yes, I am ok to fix separate problems in multiple issue, we improve step by step.

**Comment by romani** [comment]
Fix is merged
